include/kernels/blas_3d.cuh

```diff
@@ -0,0 +1,341 @@
+#pragma once
+
+#include <color_spinor_field_order.h>
+#include <index_helper.cuh>
+#include <fast_intdiv.h>
+#include <quda_matrix.h>
+#include <matrix_field.h>
+#include <kernel.h>
+//#include <kernels/contraction_helper.cuh>
+
+namespace quda
+{
+
+  struct baseArg : kernel_param<> {
+    int_fastdiv X[4]; // grid dimensions
+
+    baseArg(dim3 threads, const ColorSpinorField &x) : kernel_param(threads)
+    {
+      for (int i = 0; i < 4; i++) { X[i] = x.X()[i]; }
+      if (x.SiteSubset() == 1) X[0] = 2 * X[0];
+    }
+  };
+
+  template <typename Float, int nColor_> struct copy3dArg : baseArg {
+    using real = typename mapper<Float>::type;
+    static constexpr int nColor = nColor_;
+    static constexpr int nSpin = 1;
+    static constexpr bool spin_project = false;
+    static constexpr bool spinor_direct_load = false; // false means texture load
+
+    // Create a typename F for the ColorSpinorFields
+    typedef typename colorspinor_mapper<Float, nSpin, nColor, spin_project, spinor_direct_load>::type F;
+
+    F y;
+    F x;
+    const int slice;
+
+    copy3dArg(ColorSpinorField &y, ColorSpinorField &x, int slice) :
+      baseArg(dim3(y.VolumeCB(), y.SiteSubset(), 1), y), y(y), x(x), slice(slice)
+    {
+    }
+  };
+
+  template <typename Arg> struct copyTo3d {
```

Comment:
Should we be explicitly enforcing `x.X(3) == 1` within `copyTo3d` and `copyFrom3d`? I know there's a check in [lib/blas_3d.cu](https://github.com/lattice/quda/pull/1506/files#diff-1c272c738b087a6e48b33efe392437636acdfe9fabad2e4a1be03198a16ab4f4) , but maybe there's a straightforward way to do it here as well.

Reply:
I don't think a check here is merited.  As you note there's already a check in the code that calls this in blas_3d.cu: one can get stuck ingoing deeper into wonderland if you don't trust anything at a given level.....